{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## FindIt4Me \n",
    "\n",
    "### Overview\n",
    "FindIt4Me is a tool within a python notebook that takes a list of gene, protein, or sequence identifiers as input, then finds the corresponding identifiers in Uniprot, Ensembl, and NCBI.  Enabling teaching mode will turn on extra output, and is intended to help newbies to python understand file parsing and data wrangling.\n",
    "\n",
    "The API for Ensembl is a fancy version of the REST API, which integrates well with many workflows.  I've added comments in that section to explain the steps that I take.\n",
    "\n",
    "The API for uniprot uses python 2.7.x, so there are two ways to go from here:\n",
    "\n",
    "  1) update the python code from 2.7.x to 3.x\n",
    "\n",
    "  2) use another API and point the python notebook to run it using that API/language\n",
    "\n",
    "I've used route 1 before, and my only comment is however much time you think it will take, triple that.\n",
    "Regarding route 2, I tested the perl API, and it still is workable, meaning it did not throw any errors when I copied and pasted the example code, and then ran it (this is very good news).  The notebook will call a perl script based on that example to query uniprot, which will write the output to a temporary file in a format that I specify.  \n",
    "\n",
    "There is no one answer to situations like this, and my recommendation is to first see if there is an API that uses python, perl, or bash.  The reason is it is (fairly) simple for these scripting languages to send and receive arguments, or files as arguments.  This then allows you to run your workflow from your language of choice (like a python notebook), and then call that API as external executable command from your notebook or workflow.  You can also do the same with a java, ruby, or a C-style API, but be aware that you will have to be very mindful of I/O.  In other words, when java writes to a file, it may not be in the format that you expect or want, which can create extra work for you parsing and data wrangling.  If R is offered as an API, then I'd recommend using it as well, with the only caveat being that you will need to shape your workflow to parse tables and SQL-like data objects.\n",
    "\n",
    "### Setup\n",
    "Each input file must be a single list of identifiers. \n",
    "\n",
    "Example:\n",
    "\n",
    "        Identifier1\n",
    "        Identifier2\n",
    "        Identifier3\n",
    "        \n",
    "If you have a header, it **must** have a '#' in front to mark it as such.  \n",
    "\n",
    "Example:\n",
    "\n",
    "        #header\n",
    "        Identifier1\n",
    "        Identifier2\n",
    "        Identifier3\n",
    "     \n",
    "\n",
    "To enable teaching mode, copy and paste this line into the first line of the next cell before running it:\n",
    "\n",
    "        enableTeachingModeBool = True\n",
    "\n",
    "Important Note: To turn off teaching mode after you have turned it on, you must relaunch the notebook, or copy and paste *enableTeachingModeBool = False* in any cell after this one, and then run that cell.\n",
    "\n",
    "### Instructions\n",
    "Individual sections of the workbook can be run, or the entire notebook can be run, depending on what your need is.\n",
    "\n",
    "* Part 1 \n",
    "  * Part 1 will take as input a list of PDB protein identifiers, and an identifier that you provide to it.  The notebook then scans the Uniprot database for a matching identifier.  The default is NCBI Refseq, but others may be searched as well.   \n",
    "  * Here is the chart for what the python notebook Part 1 will retrieve.  Note that attempting to match a protein to a gene identifier may produce unexpected results:\n",
    "\n",
    "| What you want | What you should enter | What will happen |\n",
    "|:---|:---:|:---:|\n",
    "| Match Uniprot to NCBI | `P_REFSEQ_AC` | Scans Uniprot, then returns the RefSeq-Protein-ID |\n",
    "| Match Uniprot to EMBL | `EMBL_ID`  | Scans Uniprot, then returns the EMBL-ID  |\n",
    "| Match Uniprot to Ensembl | `ENSEMBL_ID` | Scans Uniprot, then returns the matching Ensembl identifier |\n",
    "| Match Uniprot to Ensembl-Protein | `ENSEMBL_PRO_ID` | Scans Uniprot, then returns the matching Ensembl protein identifier |\n",
    "| Match Uniprot to Ensembl-Transcript | `ENSEMBL_TRS_ID` | Scans Uniprot, then returns the matching Ensembl transcript identifier |\n",
    "| Match Uniprot to GeneName | `GENENAME` | Scans Uniprot, then returns the matching GeneName from Uniprot |\n",
    "\n",
    "\n",
    "  * There will be two output files.  The file beginning with 'outFile_perl' will be the raw output from the Uniprot query; you can check this to verifywhat database it queried.  The file beginning with 'uniprot_scan' (this is the file you want) will be a file with the format: \n",
    "\n",
    "          PDB_ID,Matched-ID\n",
    "\n",
    "  * Where 'Matched-ID' is the identifier that matched to what you provided.\n",
    "\n",
    "  * A full list of identifiers can be found here: https://www.uniprot.org/help/api_idmapping\n",
    "\n",
    "\n",
    "* Part 2 \n",
    "  * Part 2 will take as input a list of gene, transcript, or protein identifiers, and return a list with 3 columns.  Column 1 will be the original identifier, and will never be blank.  Column 2 will be the Ensembl identifier, and will also never be blank.  Column 3 will be the NCBI identifier if one can be found:\n",
    "\n",
    "        ID,Ensembl\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 1: Match Uniprot to NCBI\n",
    "\n",
    "In the next cell, enter the filename that contains the list of identifiers for the *inFileName* value, then copy and paste the appropriate label (see chart above) for the databse that you want to scan.  Finally, run that cell and follow the instructions in the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Ready to continue!  Select the next cell and run it.\n"
     ]
    }
   ],
   "source": [
    "inFileName = 'inFile_perl.txt'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# code\n",
    "parsed_list = []\n",
    "headerLine = ''\n",
    "\n",
    "# create timestamp so files don't overwrite each other\n",
    "import time\n",
    "import datetime\n",
    "t_stamp = time.time()\n",
    "t_stamp_string = datetime.datetime.fromtimestamp(t_stamp).strftime('%Y%m%d_%H%M%S')\n",
    "outFileStringName = 'outFile_perl' + t_stamp_string + '.txt'\n",
    "\n",
    "# file checking\n",
    "try:\n",
    "    with open(inFileName, 'r') as fOpen:\n",
    "        for i in fOpen:\n",
    "            i = i.rstrip('\\r\\n')\n",
    "            if i[0] == '#':\n",
    "                headerLine = i\n",
    "                continue\n",
    "            else:\n",
    "                parsed_list.append(i)\n",
    "except FileNotFoundError:\n",
    "    print('It does not look like you entered a file for the input_file_name value\\nor that file cannot be found.  Please check what you entered, and run this cell again.')\n",
    "\n",
    "# teaching mode (verbose output) toggle\n",
    "enableTeachingMode = False\n",
    "try:\n",
    "    if enableTeachingModeBool:\n",
    "        enableTeachingMode = enableTeachingModeBool\n",
    "except NameError:\n",
    "    enableTeachingMode = False\n",
    "    \n",
    "if enableTeachingMode:\n",
    "    print('\\n\\nTeaching mode turned on.\\n\\n')\n",
    "    \n",
    "if parsed_list:\n",
    "    print('\\nReady to continue!  Select the next cell and run it.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Select the next cell and run it\n"
     ]
    }
   ],
   "source": [
    "%%bash -s \"$inFileName\" \"$outFileStringName\"\n",
    "\n",
    "perl fetch_with_perl.pl $1 > $2\n",
    "sleep 2\n",
    "echo 'Select the next cell and run it'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Uniprot scan complete!  If you do not see any errors, continue to Part 2 to scan Ensembl.\n"
     ]
    }
   ],
   "source": [
    "l = []\n",
    "with open(outFileStringName, 'r') as fOpen:\n",
    "    for i in fOpen:\n",
    "        i = i.rstrip('\\r\\n')\n",
    "        iSplit = i.split('\\t')\n",
    "        if len(iSplit) > 1:\n",
    "            l.append((iSplit[0], iSplit[1]))\n",
    "            continue\n",
    "        l.append(i)\n",
    "\n",
    "# delete headers\n",
    "l1 = l[1:] \n",
    "l = list(filter(lambda x: x[0][0] != '#', l1))\n",
    "l = l[2:]\n",
    "# remove whitespace or tabs\n",
    "res_list = []\n",
    "for i in l:\n",
    "    if len(i) == 2:\n",
    "        s = i[0] + ',' + i[1]\n",
    "        res_list.append((i[0], i[1]))\n",
    "    else:\n",
    "        ln = i.split(' ')\n",
    "        ln_parsed = list(filter(lambda x: x != '', ln))\n",
    "        res_list.append((ln_parsed[0], ln_parsed[1]))\n",
    "\n",
    "t_stamp = time.time()\n",
    "t_stamp_string = datetime.datetime.fromtimestamp(t_stamp).strftime('%Y%m%d_%H%M%S')\n",
    "uniprot_outFile = 'uniprot_scan' + t_stamp_string + '.txt'\n",
    "uniprot_results = []\n",
    "for i in parsed_list:\n",
    "    r = list(filter(lambda x: x[0] == i, res_list))\n",
    "    if not r:\n",
    "        # print('No entry for ' + str(i))\n",
    "        uniprot_results.append((i, None))\n",
    "        with open(uniprot_outFile, 'a') as fWrite:\n",
    "            outString = i + ',' + 'None' + '\\n'\n",
    "            fWrite.write(outString)\n",
    "    else:\n",
    "        # print('Found value: ' + r[0][0] + ',' + r[0][1])\n",
    "        uniprot_results.append((r[0][0], r[0][1]))\n",
    "        with open(uniprot_outFile, 'a') as fWrite:\n",
    "            outString = r[0][0] + ',' + r[0][1] + '\\n'\n",
    "            fWrite.write(outString)   \n",
    "\n",
    "print('\\nUniprot scan complete!  If you do not see any errors, continue to Part 2 to scan Ensembl.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Part 2: Match Ensembl to NCBI\n",
    "\n",
    "In the next cell, enter the filename that contains the list of identifiers for the *inFileName* value, then run that cell and follow the instructions in the output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Ready to continue!  Select the next cell and run it.\n"
     ]
    }
   ],
   "source": [
    "inFileName = 'ens_list.txt'\n",
    "organismName = 'human' # must be the format used by Ensembl, example: human or homo_sapiens for \"human\"\n",
    "\n",
    "gName = ''\n",
    "proceedBool = True\n",
    "if not inFileName:\n",
    "    print('Please enter a name for the list of identifiers!')\n",
    "    proceedBool = False\n",
    "if not organismName:\n",
    "    print('Please enter an organism name that is used by Ensembl!')\n",
    "    proceedBool = False\n",
    "else:\n",
    "    gName = organismName\n",
    "\n",
    "if proceedBool:\n",
    "    print('\\nReady to continue!  Select the next cell and run it.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ensembl scan complete!  If you do not see any errors, continue to Part 3 to join the lists.\n"
     ]
    }
   ],
   "source": [
    "import requests, sys\n",
    "import re\n",
    "import time\n",
    "import random\n",
    "import argparse\n",
    "import hashlib\n",
    "import time\n",
    "import datetime\n",
    "\n",
    "t_stamp = time.time()\n",
    "t_stamp_string = datetime.datetime.fromtimestamp(t_stamp).strftime('%Y%m%d_%H%M%S')\n",
    "\n",
    "outFileString = 'outFile_ensembl' + t_stamp_string + '.txt'\n",
    "\n",
    "\n",
    "def lookupEnsembleID(ugeneID, ensQuery, genomeName, verboseBool):\n",
    "    if verboseBool:\n",
    "        print(\"Checking\")\n",
    "        print(ensQuery)\n",
    "    uniqueID = ugeneID\n",
    "    n = random.random() + 0.5\n",
    "    time.sleep(n)\n",
    "    server = \"https://rest.ensembl.org/\"\n",
    "    lookupIDstring = \"xrefs/symbol/\"\n",
    "    # speciesString = \"armadillo/\"\n",
    "    speciesString = genomeName + \"/\"\n",
    "    idName = ensQuery\n",
    "    ext1 = \"?content-type=application/json\"\n",
    "    q = server+lookupIDstring+speciesString+idName+ext1\n",
    "    r = requests.get(q, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    if not r.ok:\n",
    "        if verboseBool:\n",
    "            print(q)\n",
    "            print(ensQuery+\"\\t\"+\"MATCH ERROR\")\n",
    "        else:\n",
    "            print(\"Error\")\n",
    "        return (uniqueID, (ensQuery, 0))\n",
    "    else:\n",
    "        decoded = r.json()\n",
    "        if not decoded:\n",
    "            if verboseBool:\n",
    "                print(\"couldn't find match in Ensembl Lookup\")\n",
    "            return (uniqueID, (ensQuery, 0))\n",
    "        else:\n",
    "            res = decoded[0]\n",
    "            ensembleName = res[\"id\"]\n",
    "            if verboseBool:\n",
    "                print(\"Found match for\")\n",
    "                print(res)\n",
    "            return (uniqueID, (ensQuery, ensembleName))\n",
    "\n",
    "ensQuery_list = []\n",
    "with open(inFileName, 'r') as fOpen:\n",
    "    for i in fOpen:\n",
    "        if i[0] == '#':\n",
    "            continue # just skip the header for this part\n",
    "        i = i.rstrip('\\r\\n')\n",
    "        ensQuery_list.append(i)\n",
    "\n",
    "ensQuery_results = []\n",
    "for i in ensQuery_list:\n",
    "    uID = hashlib.md5(i.encode('utf-8')).hexdigest()\n",
    "    r = lookupEnsembleID(uID, i, gName, False)\n",
    "    ensQuery_results.append(r[1])\n",
    "\n",
    "for i in ensQuery_results:\n",
    "    with open(outFileString, 'a') as fWrite:\n",
    "        resString = ''\n",
    "        if not i[1]:\n",
    "            resString = 'None'\n",
    "        else:\n",
    "            resString = str(i[1])\n",
    "        s = str(i[0]) + ',' + resString + '\\n'\n",
    "        fWrite.write(s)\n",
    "        \n",
    "if ensQuery_results:\n",
    "    print('Ensembl scan complete!  If you do not see any errors, continue to Part 3 to join the lists.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
